src/libstore/local-overlay-store.cc

```diff
@@ -188,6 +188,34 @@ void LocalOverlayStore::deleteGCPath(const Path & path, uint64_t & bytesFreed)
     }
 }
 
+void LocalOverlayStore::optimiseStore()
+{
+    Activity act(*logger, actOptimiseStore);
+
+    // Note for LocalOverlayStore, queryAllValidPaths only returns paths in upper layer
+    auto paths = queryAllValidPaths();
+
+    act.progress(0, paths.size());
+
+    uint64_t done = 0;
+
+    for (auto & path : paths) {
+        if (lowerStore->isValidPath(path)) {
+            // Deduplicate store path
+            deletePath(toUpperPath(path));
+        }
+        done++;
+        act.progress(done, paths.size());
+    }
```

Comment:
Likewise, it would be nice to still hard link within the upper layer, but that is also a bit more invasive.